### PR TITLE
Add base class and one implementation of a record filter

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Add doxygen comments to filter classes
- Implemented simple epoch inference with the
  ProxyEpochRecordFilter
- Filtering done in the ApplicationSampler
- Add plumbing to add record filter to ApplicationSampler

Change-Id: Id5bc5b97d3cc2d5f5953d3580f6f403bb59452ce
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

### CODE REVIEW

File: src/ProxyEpochRecordFilter.hpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+#ifndef PROXYEPOCHRECORDFILTER_HPP_INCLUDE
+#define PROXYEPOCHRECORDFILTER_HPP_INCLUDE
+
+#include "RecordFilter.hpp"
+
+namespace geopm
+{
+    /// @brief Filter that can be used in place of the
+    ///        EpochRecordFilter to synthesize epoch events from a
```
Comment (dianarg) [inline src/ProxyEpochRecordFilter.hpp]: EpochRecordFilter is going away, right?

File: src/ProxyEpochRecordFilter.hpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    ///        sequence of region entry events.  The filter also
+    ///        passes hint events through.
```
Comment (dianarg) [inline src/ProxyEpochRecordFilter.hpp]: update to indicate all events except epoch are passed though

File: src/RecordFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#include "config.h"
+
+#include "RecordFilter.hpp"
+#include "ProxyEpochRecordFilter.hpp"
+#include "Helper.hpp"
+#include "Exception.hpp"
+
+namespace geopm
+{
+    std::unique_ptr<RecordFilter> RecordFilter::make_unique(const std::string &name)
```
Comment (dianarg) [inline src/RecordFilter.cpp:42]: please add tests for this

File: src/ApplicationSampler.cpp [hunk 4]
```diff
@@ -86,7 +92,15 @@ namespace geopm
     }
 
     ApplicationSamplerImp::ApplicationSamplerImp()
+        : ApplicationSamplerImp(environment().record_filter())
+    {
+
+    }
+
+    ApplicationSamplerImp::ApplicationSamplerImp(const std::string &filter_name)
```
Comment (dianarg) [inline src/ApplicationSampler.cpp]: it might be easier to test the passthrough if RecordFilters are injected instead of the name

**Review by dianarg — CHANGES_REQUESTED** [review]

File: src/ApplicationSampler.cpp (hunk 4 above)
Comment (cmcantalupo) [reply src/ApplicationSampler.cpp]: We need a way to generate record filters for each process after construction.  That is facilitated with a string now.  I'm not sure how we want to mock out the RecordFilter::make_unique() call, which is where we will run into testing problems.

Comment (cmcantalupo) [reply src/ApplicationSampler.cpp]: maybe:

```
ApplicationSamplerImp(std::function<std::unique_ptr<RecordFilter>()> filter_factory);
```

File: test/RecordFilterTest.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+TEST_F(RecordFilterTest, invalid_filter_name)
+{
+    GEOPM_EXPECT_THROW_MESSAGE(RecordFilter::make_unique("invalid_filter_name"),
+                               GEOPM_ERROR_INVALID,"parse name");
+}
+
+TEST_F(RecordFilterTest, make_proxy_epoch)
+{
+    std::shared_ptr<RecordFilter> filter = RecordFilter::make_unique("proxy_epoch,0xabcd1234");
+    EXPECT_TRUE(filter);
```
Comment (dianarg) [inline test/RecordFilterTest.cpp]: maybe EXPECT_NO_THROW would be clearer?

**Review by dianarg — DISMISSED** [review]

**Review by dianarg — APPROVED** [review]

**Pull request merged**